# Description: remove upstream PATH/LD_LIBRARY_PATH workaround
--- a/swig/python/__init__.py
+++ b/swig/python/__init__.py
@@ -29,17 +29,6 @@ import numpy as np
 
 import ctypes
 
-# add to PATH to make dlopen find the libraries
-if "PATH" in os.environ:
-  os.environ["PATH"] = os.path.dirname(__file__) + os.pathsep +  os.environ["PATH"]
-else:
-  os.environ["PATH"] = os.path.dirname(__file__)
-
-if "LD_LIBRARY_PATH" in os.environ:
-  os.environ["LD_LIBRARY_PATH"] = os.path.dirname(__file__) + os.pathsep + os.environ["LD_LIBRARY_PATH"]
-else:
-  os.environ["LD_LIBRARY_PATH"] = os.path.dirname(__file__)
-
 from casadi_loader import *    # import everything
 import casadi_loader as casadi # import everything
